src/repr/type_random.ml

```diff
@@ -0,0 +1,138 @@
+open Type_core
+open Staging
+module R = Random.State
+
+type 'a random = (R.t -> 'a) staged
+
+let ( let+ ) x f =
+  let x = unstage x in
+  stage (f x)
+
+let ( and+ ) a b = stage (unstage a, unstage b)
+let return x = stage (fun _ -> x)
+
+(* Sample lengths according to a geometric distribution by inverse
+   transform sampling an exponential one (characterised by [mean_length = λ⁻¹])
+   and rounding to integer values. *)
+let pick_len : mean:int -> len -> R.t -> int =
+  let bound len x =
+    match len with
+    | `Int | `Int64 -> x
+    | `Int8 -> min x ((1 lsl 8) - 1)
+    | `Int16 -> min x ((1 lsl 16) - 1)
+    | `Int32 -> min x ((0x7fff lsl 16) lor 0xffff)
+  in
+  fun ~mean l s ->
+    match l with
+    | (`Int | `Int8 | `Int16 | `Int32 | `Int64) as l ->
+        bound l (Float.to_int (-.Float.log (R.float s 1.) *. Float.of_int mean))
+    | `Fixed i -> i
+
+let indexable :
+    type a b.
+    mean_len:int -> len -> (int -> (int -> a) -> b) -> a random -> b random =
+ fun ~mean_len len init elt ->
+  let+ elt = elt in
+  fun s -> init (pick_len ~mean:mean_len len s) (fun _ -> elt s)
+
+module Record_deriver = Fields_folder (struct
+  type ('a, 'b) t = R.t -> 'b -> 'a
+end)
+
+let int32 =
+  let open Int32 in
+  let bits s = of_int (R.bits s) in
+  fun s -> logxor (bits s) (shift_left (bits s) 30)
+
+let int64 =
+  let open Int64 in
+  let bits s = of_int (R.bits s) in
+  fun s ->
+    logxor (bits s) (logxor (shift_left (bits s) 30) (shift_left (bits s) 60))
```

Comment:
Is there a difference if you write it like this:

```suggestion
let bits of_int s = of_int (R.bits s)

let int32 s =
  let open Int32 in
  logxor (bits of_int s) (shift_left (bits of_int s) 30)

let int64 s =
  let open Int64 in
  logxor (bits of_int s)
    (logxor (shift_left (bits of_int s) 30) (shift_left (bits of_int s) 60))
```

Reply:
There's no difference if `bits` is inlined (which I suspect Flambda will do), but I'm not sure it's easier to read.